### DISCUSSION TITLE
Test & folder structure for multiple server

### DISCUSSION THREAD

**Discussion by Goopil** [TriPSs/nestjs-query#42] [Q&A]
I was starting to fiddle a bit with fastify plateforme with (ground work to support mercurius, but let's go slowly about this)

And found myself duplicating each example contexte with the exact same test in it. Not sur if this is the proper way. Only the before and after each changes.

What is your pick on this ?
(So fare all green by the way)

**Comment by TriPSs** [comment]
Yea I had the same feeling about this when updating the project, think it's a good idea to do the E2E testing differently. The original examples could also be run independently and even though the idea behind that is awesome I don't think a lot off ppl do this. 

Some things I think we could do:
- Merge examples more together so duplicated code becomes less.

Off topic regarding mercurius: I'm now running this project btw with no problems with nestjs + mercurius :)

**Reply by TriPSs** [reply]
We could also just completely separate the E2E testing of the examples.

**Comment by Goopil** [comment]
Well you already did the main tedious job :) splitting updating and migrating to a more monorepo package manager. 

We could leverage this ? What matter is to not loose the substance of what has been tested (the initial package owner did an amazing job.) But it's true that this dos not help in context reproduction (as of now)

What could help is having server context splitted in package (only the mains.ts and app.ts would changes) and splitting the server init from the tests. (Well having real e2e at the end of the day.)

Is this proposal to your liking?

**Reply by TriPSs** [reply]
Now that i'm checking the tests again I see that the `main.ts` are actually never used anymore 😅 

I agree it's a good idea to split the init from the tests :)

**Comment by Goopil** [comment]
Congratz for the mercurius integration  :) did you see some perf improvement? (Has it should)

Did you have any problem with errors response format ?

**Reply by TriPSs** [reply]
The API is not live yet soo not quite sure ;)

Yes and no with the errors: I almost always have custom error exceptions and I did create a custom exception filter as all my errors now where always logged to the console, looks like this:

```ts
// This is ofc inside the bootstrap
app.useGlobalFilters(new GlobalExceptionFilter())


import { Catch, ExceptionFilter } from '@nestjs/common'
import mercurius from 'mercurius'

interface ExtensionWithCode {
  code?: string
}

@Catch(mercurius.ErrorWithProps)
export class GlobalExceptionFilter implements ExceptionFilter {

  private hasCode(exception: ExtensionWithCode) {
    return exception?.code && exception?.code.startsWith('api.')
  }

  public catch(exception: mercurius.ErrorWithProps) {
    // Returning a true value will prevent the original logger from logging
    return this.hasCode(exception as ExtensionWithCode)
  }

}
```

My custom errors all extend `mercurius.ErrorWithProps` and look like this:
```ts
import { HttpStatus } from '@nestjs/common'
import mercurius from 'mercurius'

export class ItemNotFoundException extends mercurius.ErrorWithProps {

  constructor(message = 'Recourse could not be found!') {
    super(message, {
      message,
      code: 'api.recourse-not-found'
    }, HttpStatus.NOT_FOUND)
  }

}
```

**Reply by Goopil** [reply]
Ok thanks for sharing !

Beware of the validation exceptions. If I remember correctly, they are swallowed in a httpExeption loosing the original message, context and format....

**Reply by TriPSs** [reply]
Good point, I have a custom one for that to:
```ts
app.useGlobalPipes(
    new ValidationPipe({
      exceptionFactory: ValidationException.exceptionFactory
    })
  )
```

And the `ValidationException`
```ts
import { HttpStatus } from '@nestjs/common'
import mercurius from 'mercurius'

import type { ValidationError } from 'class-validator'

type ValidationFields = Record<string, unknown>

export class ValidationException extends mercurius.ErrorWithProps {

  public static MESSAGE = 'Validations failed!'

  constructor(message = ValidationException.MESSAGE, fields: ValidationFields = {}) {
    super(message, {
      code: 'api.validation',
      fields
    }, HttpStatus.BAD_REQUEST)
  }

  public static exceptionFactory(errors: ValidationError[]) {
    return new ValidationException(
      ValidationException.MESSAGE,
      ValidationException.getFieldsFromErrors(errors)
    )
  }

  public static getFieldsFromErrors(errors: ValidationError[]): ValidationFields {
    return errors.reduce((userInputError, validationError) => {
      if (!validationError.constraints) {
        if (validationError.children) {
          return {
            ...userInputError,
            ...this.getFieldsFromErrors(validationError.children)
          }
        }

        return userInputError
      }

      const firstErrorConstraint = Object.keys(validationError.constraints)[0]
      userInputError[validationError.property] = validationError.constraints[firstErrorConstraint]

      return userInputError
    }, {})
  }

}

```

**Reply by Goopil** [reply]
i forgot to say thanks for the examples 🚀
